File: Chapter 10/Python/section_10_3_resilience_decorators/resilient/infrastructure/adapters/local_queue_adapter.py
import persistqueue
from ...core.ports.message_queue import MessageQueue

class LocalQueueAdapter(MessageQueue):
    """
    PHYSICAL INFRASTRUCTURE ADAPTER (Zero-Server Local Queue)
    
    @description 
    This adapter uses a disk-backed SQLite queue to simulate a decoupled 
    message broker. It provides durability without requiring external 
    accounts, API keys, or server installations.
    
    PRODUCTION ARCHITECTURE (How this works on a real server):
    In a live production environment (e.g., using RabbitMQ, AWS SQS, or Kafka), 
    this 'enqueue' method performs a 'Fire and Forget' handoff:
    
    1. THE PHYSICAL CONNECTION: Instead of writing to a local file, the 
       adapter maintains a TCP/IP connection (often persistent) to a 
       Message Broker cluster.
       
    2. SERIALIZATION & PROTOCOL: The Python dictionary is serialized 
       (typically to JSON or Protobuf) and wrapped in a protocol frame 
       (like AMQP or MQTT).
       
    3. THE HANDOFF (TEMPORAL DECOUPLING): The 'enqueue' call is non-blocking 
       from the Core's perspective. Once the Broker sends a 'Commit' or 'Ack' 
       confirming the message is safely in its own storage, the Application 
       is free to move on.
       
    4. WORKER INDEPENDENCE: A separate 'Consumer' or 'Worker' service 
       (potentially on a different server or auto-scaling cluster) polls 
       the Broker and eventually processes the payment. 
       
    *CRITICAL LESSON*: The Core doesn't care if the 'Worker' is alive or 
    dead at the moment of 'enqueue'. This is how we survive 
    infrastructure spikes and vendor downtime.
    """
    def __init__(self, path="./queue_storage"):
        # Creates a physical SQLite database on the local disk.
        # This acts as our "Message Broker" for the lab.
        self.queue = persistqueue.SQLiteQueue(path, auto_commit=True)

    def enqueue(self, payload: dict):
        """
        Physically persists the transaction data to the local disk.
        """
        # This is a synchronous write to the local database file.
        # It mimics the 'send' operation to a production message server.
        self.queue.put(payload)
        order_id = payload.get("order_id", "UNKNOWN")
        
        print(f"      [Local Queue] DATA PERSISTED: Order {order_id} secured in ./queue_storage/")

File: Chapter 10/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter10/section_10_3_resilience_decorators/resilient/infrastructure/adapters/LocalQueueAdapter.java
package com.grokkingsoftwarearchitecture.chapter10.section_10_3_resilience_decorators.resilient.infrastructure.adapters;

import com.grokkingsoftwarearchitecture.chapter10.section_10_3_resilience_decorators.resilient.core.ports.MessageQueue;
import org.h2.mvstore.MVMap;
import org.h2.mvstore.MVStore;
import java.util.Map;
import java.util.UUID;

/**
 * PHYSICAL INFRASTRUCTURE ADAPTER (Zero-Server Local Queue):
 * * @description 
 * This adapter uses H2's MVStore to create a disk-backed queue. It provides 
 * durability for the laboratory environment without requiring external brokers 
 * or cloud services.
 * * PRODUCTION ARCHITECTURE (How this works on a real server):
 * In a live production environment (e.g., using RabbitMQ or AWS SQS), 
 * this 'enqueue' method functions as follows:
 * * 1. THE CONNECTION: Instead of a local file path, the adapter opens a 
 * socket connection to the broker cluster.
 * 2. SERIALIZATION: The 'payload' Map is serialized into JSON or a 
 * binary format like Protobuf to be sent over the wire.
 * 3. ACKNOWLEDGMENT: The server sends back an 'Ack' once the message 
 * is persisted to its own distributed storage.
 * 4. ISOLATION: The Application Core is released immediately, while 
 * a separate Worker process eventually pulls the message.
 */
public class LocalQueueAdapter implements MessageQueue {
    private final String dbPath;

    public LocalQueueAdapter(String dbPath) {
        this.dbPath = dbPath;
    }

    @Override
    public void enqueue(Map<String, Object> payload) {
        // Open the physical store
        try (MVStore s = new MVStore.Builder().fileName(dbPath).open()) {
            MVMap<String, Map<String, Object>> map = s.openMap("payment_queue");
            
            // Use a random key for the queue entry
            map.put(UUID.randomUUID().toString(), payload);
            
            // Physically flush to disk
            s.commit();
        }

        System.out.println("      [Local Queue] DATA PERSISTED: Message secured in " + dbPath);
    }
}